File: ios/App/View/Navigation/NavigationShellView.swift
import SwiftUI

/// The application shell: a top bar with a menu button plus a slide-over
/// sidebar drawer over the content — the native twin of the web
/// design-system AppShell (see docs/shell.md). Generic over the shell nav
/// schema (ShellNavModels.swift); apps bind it with their nav config and a
/// content host, like KernelShellView does for the Identity exemplar.
struct NavigationShellView<Content: View>: View {
  @Environment(\.uiThemeTokens) private var theme

  let title: String
  let sections: [ShellNavSection]
  var drillUp: ShellDrillUp? = nil
  var profile: ShellProfile? = nil
  @Binding var selectedItemId: String
  @ViewBuilder let content: (String) -> Content

  @State private var isSidebarOpen = false

  private static var drawerWidth: CGFloat { 300 }

  var body: some View {
    ZStack(alignment: .leading) {
      VStack(spacing: 0) {
        topBar
        content(selectedItemId)
          .frame(maxWidth: .infinity, maxHeight: .infinity)
      }
      .background(theme.colors.appBg)

      if isSidebarOpen {
        theme.colors.overlayBackdrop
          .ignoresSafeArea()
          .onTapGesture { closeSidebar() }
          .transition(.opacity)
      }

      SidebarMenuView(
        title: title,
        sections: sections,
        drillUp: drillUp.map { drillUp in
          ShellDrillUp(label: drillUp.label) {
            closeSidebar()
            drillUp.action()
          }
        },
        profile: profile.map { profile in
          // Menu-item actions typically navigate, so close the drawer first.
          var wrapped = profile
          wrapped.items = profile.items.map { item in
            ShellProfileMenuItem(id: item.id, label: item.label, systemImage: item.systemImage) {
              closeSidebar()
              item.action()
            }
          }
          return wrapped
        },
        selectedItemId: selectedItemId,
        onSelect: { item in
          selectedItemId = item.id
          closeSidebar()
        }
      )
      .frame(width: Self.drawerWidth)
      .offset(x: isSidebarOpen ? 0 : -(Self.drawerWidth + 24))
      .ignoresSafeArea(edges: .bottom)
    }
    .animation(.easeInOut(duration: 0.2), value: isSidebarOpen)
    .background(hotkeyHandlers)
  }

  private var topBar: some View {
    HStack(spacing: theme.spacing.md) {
      Button {
        withAnimation(.easeInOut(duration: 0.2)) {
          isSidebarOpen.toggle()
        }
      } label: {
        Image(systemName: "line.3.horizontal")
          .font(.system(size: theme.typography.sizes.lg, weight: .medium))
          .foregroundStyle(theme.colors.textPrimary)
          .frame(width: 36, height: 36)
          .contentShape(Rectangle())
      }
      .buttonStyle(.plain)
      .accessibilityLabel("Open navigation")

      Text(activeItemLabel)
        .font(theme.typography.font(size: theme.typography.sizes.md, weight: .semibold))
        .foregroundStyle(theme.colors.textPrimary)
        .lineLimit(1)

      Spacer(minLength: 0)
    }
    .padding(.horizontal, theme.spacing.sm)
    .frame(height: 48)
    .background(theme.colors.surface)
    .overlay(alignment: .bottom) {
      Rectangle()
        .fill(theme.colors.border)
        .frame(height: 1)
    }
  }

  private var activeItemLabel: String {
    ShellNavigation.flattenItems(sections)
      .first { $0.id == selectedItemId }?
      .label ?? title
  }

  /// Cmd+Shift+<key> shortcuts from the nav schema, for hardware keyboards.
  @ViewBuilder
  private var hotkeyHandlers: some View {
    ForEach(Array(ShellNavigation.hotkeyMap(sections)), id: \.key) { key, itemId in
      if let character = key.first {
        Button("") {
          selectedItemId = itemId
          closeSidebar()
        }
        .keyboardShortcut(KeyEquivalent(character), modifiers: [.command, .shift])
        .hidden()
      }
    }
  }

  private func closeSidebar() {
    withAnimation(.easeInOut(duration: 0.2)) {
      isSidebarOpen = false
    }
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/view/navigation/NavigationShellView.kt
package com.baseapp.android.view.navigation

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.baseapp.android.components.shell.ShellNavSection
import com.baseapp.android.components.shell.ShellNavigation
import com.baseapp.android.view.theme.LocalUiTheme

/**
 * The application shell: a top bar with a menu button plus a slide-over
 * sidebar drawer over the content — the native twin of the web design-system
 * AppShell (see docs/shell.md). Generic over the shell nav schema
 * (ShellNavModels.kt); apps bind it with their nav config and a content
 * host, like KernelShellView does for the Identity exemplar.
 */
@Composable
fun NavigationShellView(
    title: String,
    sections: List<ShellNavSection>,
    selectedItemId: String,
    onSelect: (String) -> Unit,
    drillUp: ShellDrillUp? = null,
    profile: ShellProfile? = null,
    iconForItemId: Map<String, ImageVector> = emptyMap(),
    content: @Composable (String) -> Unit,
) {
    val theme = LocalUiTheme.current
    var isSidebarOpen by remember { mutableStateOf(false) }
    val drawerWidth = 300.dp
    val drawerOffset by animateDpAsState(
        targetValue = if (isSidebarOpen) 0.dp else -(drawerWidth + 24.dp),
        animationSpec = tween(durationMillis = 200),
        label = "drawerOffset",
    )

    Box(modifier = Modifier.fillMaxSize()) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .background(theme.colors.appBg),
        ) {
            TopBar(
                label = ShellNavigation.flattenItems(sections)
                    .firstOrNull { it.id == selectedItemId }
                    ?.label ?: title,
                onMenuClick = { isSidebarOpen = true },
            )
            Box(modifier = Modifier.weight(1f)) {
                content(selectedItemId)
            }
        }

        AnimatedVisibility(visible = isSidebarOpen, enter = fadeIn(), exit = fadeOut()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(theme.colors.overlayBackdrop)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                    ) { isSidebarOpen = false },
            )
        }

        Box(
            modifier = Modifier
                .fillMaxHeight()
                .width(drawerWidth)
                .offset(x = drawerOffset),
        ) {
            SidebarMenuView(
                title = title,
                sections = sections,
                selectedItemId = selectedItemId,
                onSelect = { item ->
                    isSidebarOpen = false
                    onSelect(item.id)
                },
                drillUp = drillUp?.let { drillUpConfig ->
                    ShellDrillUp(drillUpConfig.label) {
                        isSidebarOpen = false
                        drillUpConfig.action()
                    }
                },
                profile = profile?.let { profileConfig ->
                    // Menu-item actions typically navigate, so close the drawer first.
                    profileConfig.copy(
                        items = profileConfig.items.map { item ->
                            item.copy(action = {
                                isSidebarOpen = false
                                item.action()
                            })
                        },
                    )
                },
                iconForItemId = iconForItemId,
            )
        }
    }
}

@Composable
private fun TopBar(label: String, onMenuClick: () -> Unit) {
    val theme = LocalUiTheme.current
    Column(modifier = Modifier.background(theme.colors.surface)) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxWidth()
                .statusBarsPadding()
                .height(48.dp)
                .padding(horizontal = theme.spacing.sm),
        ) {
            Icon(
                imageVector = Icons.Filled.Menu,
                contentDescription = "Open navigation",
                tint = theme.colors.textPrimary,
                modifier = Modifier
                    .size(36.dp)
                    .clickable(onClick = onMenuClick)
                    .padding(6.dp),
            )
            Text(
                text = label,
                color = theme.colors.textPrimary,
                fontSize = theme.typography.sizes.md,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
                modifier = Modifier.padding(start = theme.spacing.sm),
            )
        }
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(1.dp)
                .background(theme.colors.border),
        )
    }
}
